internal/vm/vm.go

```diff
@@ -0,0 +1,215 @@
+	State() State
+
+	// Start will power on the Utility VM and put it into a running state. This will boot the guest OS and start all of the
+	// devices configured on the machine.
+	Start(ctx context.Context) error
+
+	// Stop will shutdown the Utility VM and place it into a terminated state.
+	Stop(ctx context.Context) error
+
+	// Pause will place the Utility VM into a paused state. The guest OS will be halted and any devices will have be in a
+	// a suspended state. Save can be used to snapshot the current state of the virtual machine, and Resume can be used to
+	// place the virtual machine back into a running state.
+	Pause(ctx context.Context) error
+
+	// Resume will put a previously paused Utility VM back into a running state. The guest OS will resume operation from the point
+	// in time it was paused and all devices should be un-suspended.
+	Resume(ctx context.Context) error
+
+	// Save will snapshot the state of the Utility VM at the point in time when the VM was paused.
+	Save(ctx context.Context) error
+
+	// Wait synchronously waits for the Utility VM to shutdown or terminate. A call to stop will trigger this
+	// to unblock.
+	Wait() error
+
+	// Stats returns statistics about the Utility VM. This includes things like assigned memory, available memory,
+	// processor runtime etc.
+	Stats(ctx context.Context) (*stats.VirtualMachineStatistics, error)
+
+	// Supported returns if the virt stack supports a given operation on a resource.
+	Supported(resource Resource, operation ResourceOperation) bool
+
+	// ExitError will return any error if the Utility VM exited unexpectedly, or if the Utility VM experienced an
+	// error after Wait returned, it will return the wait error.
+	ExitError() error
+}
+
+// Resource refers to the type of a resource on a Utility VM.
+type Resource uint8
+
+const (
+	VPMem = iota
+	SCSI
+	Network
+	VSMB
+	PCI
+	Plan9
+	CPUGroup
+)
+
+// Operation refers to the type of operation to perform on a given resource.
+type ResourceOperation uint8
+
+const (
+	Add ResourceOperation = iota
+	Remove
+	Update
+)
+
+// GuestOS signifies the guest operating system that a Utility VM will be running.
+type GuestOS string
+
+const (
+	Windows GuestOS = "windows"
+	Linux   GuestOS = "linux"
+)
+
+var ErrUnknownGuestOS = errors.New("unknown guest operating system supplied")
+
+// State signifies the states that a Utility VM can be in. The state of the Utility VM should be the source of truth for what
+// operations can be performed at a given moment.
+type State uint8
+
+const (
+	StatePreCreated State = iota
+	StateCreated
+	StateRunning
+	StateTerminated
+	StatePaused
+)
+
+// SCSIDiskType refers to the disk type of the scsi device. This is either a vhd, vhdx, or a physical disk.
+type SCSIDiskType uint8
+
+const (
+	SCSIDiskTypeVHD1 SCSIDiskType = iota
+	SCSIDiskTypeVHDX
+	SCSIDiskTypePassThrough
+)
+
+// SCSIManager manages adding and removing SCSI devices for a Utility VM.
+type SCSIManager interface {
```

Comment:
My thinking was that each interface would have its owned FooSupported method, rather than have a single Supported method and having to keep that in sync. But I can see how this will be fairly painful to implement.

@kevpar, what do you think?

Reply:
That's what I'd talked about before. I find it much easier to add a new const and an extra switch case in one method than add Supported calls to everything

Reply:
That's my reasoning